### DISCUSSION TITLE
Add page numbering (to remove it) to typst template

### DISCUSSION THREAD

**Discussion by Disonantemus** [jgm/pandoc#10370] [Ideas]
There are _many_ cases when page numbering is not needed (ex: one page pdf), now is enabled by default. Right now, there is no way to disable page numbering by template, as you can see, is not there (gives no output):

~~~ocaml
$ pandoc -D typst | grep " numbering:"

~~~

in.md (sample file):
~~~md
## Title

1. list item
1. list item
~~~

Checking typst output, **_is there_** (but **not** in the template):
~~~ocaml
$ pandoc in.md -s -t typst | grep " numbering:"
    numbering: "1",
~~~

---

roff has page numbering in template and I can create a custom one like this:

~~~ocaml
$ pandoc -D ms | sed -E 's/^(\.ds CF) %$/\1/g' > ms-custom.template
$ pandoc in.md -s -t ms --template=ms-custom.template -o out.pdf
~~~

but with typst, I can't, because is not in the template (is in another place).

---

### Workaround 1: modify typst output and compile

~~~ocaml
$ pandoc in.md -s -t typst | sed -E 's/^( +numbering: )"1",/\1none,/g' > out.typ
$ typst compile out.typ && rm -f out.typ
$ xdg-open out.pdf
~~~
- not ideal, because should be in the template

### Workaround 2: disable footer by template

~~~ocaml
$ pandoc -s -D typst | sed '1i #set page(footer: none)' > typst-custom.template
$ pandoc in.md --pdf-engine typst --template=typst-custom.template -o out.pdf
$ xdg-open out.pdf
~~~
- works, but not ideal, because footer can be useful

### Workaround 3: use +raw_attribute

~~~md
`#set page(numbering: none)`{=typst}

## Title

1. list item
1. list item
~~~
- only 2nd page is ok
- bad idea to edit every md file

---

### Extra: horizontalrule

The same happens with **horizontalrule** definition, right now is too small for me, I would like to change it in my custom template, and not modify typst output and compile:

~~~ocaml
$ pandoc in.md -s -t typst | grep -A2 " hor"
#let horizontalrule = [
  #line(start: (25%,0%), end: (75%,0%))
]
~~~

**Comment by jgm** [comment]
It is in the template.
Note that `default.typst` uses a partial called `template.typst`:
```
$if(template)$
#import "$template$": conf
$else$
$template.typst()$
$endif$
```
So, it should work to do this: copy `default.typst` and `template.typst` to your working directory, change `template.typst`, and use `--template=default.typst`

Point taken, though: there should be a variable controlling this.

**Reply by jgm** [reply]
Adding a `page-numbering` variable.

**Comment by Disonantemus** [comment]
It works!
I didn't know default.typst file!

## Steps to reproduce

~~~ocaml
$ mkdir 0
mkdir: created directory '0'
$ cp a.md 0/
1 folder(s) (  4.0 KiB) copied in  0.0s (  0.0 KiB/s).
$ cd 0
$ wget --quiet https://github.com/jgm/pandoc/raw/refs/heads/main/data/templates/template.typst
$ wget --quiet https://github.com/jgm/pandoc/raw/refs/heads/main/data/templates/default.typst
$ sed -i -E 's/^( +pagenumbering: )"1",/\1none,/g' template.typst
$ pandoc a.md -t typst --template=default.typst -o o.pdf
$ xdg-open o.pdf
~~~

---

It also works to have the 3 template files related to _typst_ in a different folder and use it like this:

From: /data/templates:
~~~ocaml
template.typst
default.typst
definitions.typst
~~~

~~~ocaml
$ pandoc a.md -t typst --template=/home/user/templates/default.typst -o out.pdf
~~~

---

- Shouldn't the content of the definitions.typst file be included in default.typst?

- There are only 2 definitions (hotizontalrule, endnote), and typst is the only format that has a definition file in that folder.

- The same goes for the template.typst file, doesn't it make more sense to merge the three? since that is also the only template apart from default.typst.

**Reply by jgm** [reply]
Yes, we could consider not using partials here.  I don't really remember why I did -- perhaps because the standard typst examples have a separate template file.  And there used to be more definitions.
@tarleb any thoughts on merging these into default.typst?

**Reply by iandol** [reply]
I think Typst lends itself well to separating the typst layout code from the "document" (https://typst.app/docs/tutorial/making-a-template/), which in this case is the pandoc default template. I do agree that definitions.typst is not so essential, and could be rolled into either of the other files. But as someone who has customised Typst templates for my needs, the default.typst (where pandoc variables are expanded and pandoc structural elements live) + template.typst (where Typst code builds layout) split makes sense to me.

**Reply by jgm** [reply]
I will at least incorporate definitions.typst. And in fact one of the two definitions it contains is not even necessary any more.

**Comment by iandol** [comment]
The OP could also have used header-includes metadata to inject the Typst code, no need to mess with the Typst templates at all:

``````yaml
header-includes: |
  ```{=typst}
  #set page(numbering: none)
  ```
``````

**Reply by Disonantemus** [reply]
> The OP could also have used header-includes metadata to inject the Typst code, no need to mess with the Typst templates at all:
> 
> ```yaml
> header-includes: |
>   ```{=typst}
>   #set page(numbering: none)
>   ```
> ```

I did try your code, but didn't work for me ...

_______________________________________________________________________

## header-includes inside [yaml_metadata_block](https://pandoc.org/MANUAL.html#extension-yaml_metadata_block)

I changed my example to include your code:

````ocaml
$ cat in.md
---
title: Odyssey
author: Homer
header-includes: |
  ```{=typst}
  #set page(numbering: none)
  ```
...

## Title

1. list item
1. list item
````
Previously, I did check with an online YAML validator.

Then I did check `typst` output, seems parsing is ok, with the command:

```ocaml
$ pandoc in.md -s -t typst | grep -v section | grep "numbering:"
    numbering: "1",
#set page(numbering: none)
```

But when I did generate output to pdf, with this:

```ocaml
$ pandoc in.md -s -t typst -o out.pdf ; xdg-open out.pdf
```

Page Number is still there (not removed)

- What am I doing wrong?
- When checking the output of Typst, the page number was previously defined by PanDoc.

## header-includes using [--include-in-header](https://pandoc.org/MANUAL.html#variables-set-automatically)

I created nopnt.md:
~~~ocaml
$ cat nopnt.md
```{=typst}
#set page(numbering: none)
```
~~~

I did use first example (in.md):
~~~ocaml
$ cat in.md
## Title

1. list item
1. list item
~~~

Then created a pdf using `--include-in-header`:
~~~ocaml
$ pandoc in.md -s -t typst --include-in-header=nopnt.md -o o.pdf ; xdg-open o.pdf
~~~
But the output was bad:
- First page (no page number) and this text (that shouldn't be there):
~~~
{=typst}
#set page(numbering: none)
~~~
- Second page: expected result, but with page number 2.

> _I think am using this wrong!_

**Reply by iandol** [reply]
Right, the problem here is that `header-includes` injects the `#set` command **_before_** the `#show: doc` template command:

```typst
#set page(numbering: none)

#show: doc => conf(
  title: [Odyssey],
  authors: (
    ( name: [Homer],
      affiliation: "",
      email: "" ),
    ),
  cols: 1,
  doc,
)
```

The template ALSO sets page option and so the one injected by header-includes is now lost. So my suggestion indeed doesn't work 😢 

Note: there is also an `include-before` metadata which is injected **AFTER** the `#show: doc => conf` so this would not be overwritten by the template. BUT this would introduce a new problem: every time you use a `#set page()` command a **page break** is introduced. Therefore you would see a first page with page numbers and subsequent pages without:

``````markdown
---
title: Odyssey
author: Homer
include-before: |
  ```{=typst}
  #set page(numbering: none)
  ```
...

## Title

1. list item
1. list item
``````

```typst
#show: doc => conf(
  title: [Odyssey],
  authors: (
    ( name: [Homer],
      affiliation: "",
      email: "" ),
    ),
  cols: 1,
  doc,
)

#set page(numbering: none)


== Title
<title>
+ list item
+ list item
```


<img width="666" alt="image" src="https://github.com/user-attachments/assets/794a80dd-8756-42ad-a528-819948730c7c">

Typst is quite fussy when setting page options -- so in this case changing the template to add a `page-numbering` variable is the best solution, thanks to @jgm.
